_rules/1520.md

```diff
@@ -1,21 +1,28 @@
 ---
 rule_id: 1520
 rule_category: maintainability
-title: Only use `var` when the type is very obvious
+title: Use `var` (implicit typing) and emphasize business purpose over mechanics
 severity: 1
 ---
-Only use `var` as the result of a LINQ query, or if the type is very obvious from the same statement and using it would improve readability. So don't
+In almost all cases we recommend the use of var, to emphasizes the semantic "business purpose" of the variable and downplay the "mechanical" details of its storage (matching Domain Driven Design ubiquitous language principles).
 
-	var item = 3;                              // what type? int? uint? float?
-	var myfoo = MyFactoryMethod.Create("arg"); // Not obvious what base-class or			
-	                                           // interface to expect. Also
-	                                           // difficult to refactor if you can't
-	                                           // search for the class
+    var amount = 2.1m;
+    var referral = new Referral("jane.smith@acme.com");
+    var discount = discountService.GetDiscountFromReferral(referral)
+    var discountedAmount = discount.ApplyTo(amount);
```

Comment:
Hmm, `discount` is not clear enough for me. Is it a percentage? Is it an absolute value? And if so, what unit?
Same for `discountedAmount`. What's the unit here? 2 cows? 2 euros? A percentage?

Reply:
This is where context comes in. In general you would expect an entity here. If it was a percentage or an amount, that should be added to the method and the variable name.

Same for dfiscountedAmount and since amount is passed it should be clear to be the same kind of unit.

I could get into much more detail but I'm not sure this lemma is the right place or if I should write a separate blog and refer to it?

Reply:
> In general you would expect an entity here.

That would have never crossed my mind. 

> If it was a percentage or an amount, that should be added to the method and the variable name.

Yes, but any amount has a _unit_. That's why we have types. And that's why an association between classes in UML have a name _and_ a role.

Reply:
My reply would be: wvery variable has a _context_. A semantic one. I'm not sure what your comment about UML refers to. What kind of association do you mean? Parent-child? Entity-to-entity? Are we still caring about ubiquitous language here? I'm expecting a domain driven design approach here, but I'm clearly missing something.

Reply:
The mere fact that you have to explain all of that, reinforces my opinion that in most cases, the type _is_ important. 

My point about relationships in UML, is that in a class diagram, the end of a relationship has two important attributes: the type of the class it points to and a name that decribes the role of that type in a relationship. A variable, class field or return value  _are_ examples of such relationship and thus have a type and a role. In your example, the name of the role would be the same as the type, but I would still show this explicitly, so the reviewer doesn't have to wonder what happens there. And for the record, I like the approach you take by encapsulating discounts as an entity. I'm a big fan (and quite experienced) in DDD.

Reply:
Looking back at this after two years, I can easily stand by what was said previously. The functional intent is all that matters, and the type of question you have for us is exclusively a matter of naming convention. 

Explicit variables being necessary is just really the exception to the rule. 

And in our case, a very rare exception indeed. Rare cases where using a Task.FromResult in a test for mocking an async return value wants a specific interface for a list. In this case you could also use something else, but precisely because we use var everywhere, not using it in that location tells everyone immediately that there the type *did* matter.

So we'll have to agree to disagree then. But I'm willing to bet the worldwide trend will see var becoming the norm (as I think is already happening, if you look at the defaults from various rulesets).